### PR TITLE
Api Implementation

### PR DESCRIPTION
**Pull request by nigelmann**
- Implementation of controllers
- Implementation of services
- Implementation of mappers

I have added `FIXME` and `TODO` where attention is due

I think for now this can be merged and those points fixed later.

Some modification in the `api.yaml` file will be necessary.

issue: #26

### CODE REVIEW

File: apps/backend/src/main/java/ch/heig/pdg/backend/BackendApplication.java [hunk 1]
```diff
@@ -23,6 +24,11 @@ public FilterRegistrationBean<AuthorizationFilter> filterRegistrationBean() {
         return registrationBean;
     }
 
+    @Bean
```
Comment (noahboegli) [inline apps/backend/src/main/java/ch/heig/pdg/backend/BackendApplication.java:27]: You like beans ? 

https://youtu.be/E-GD1_0MUGw?t=119

File: apps/backend/src/main/java/ch/heig/pdg/backend/dto/mapping/AbstractDataMapper.java
Comment (noahboegli) [inline apps/backend/src/main/java/ch/heig/pdg/backend/dto/mapping/AbstractDataMapper.java:1]: It's great

But let's say I don't fancy having the repository doing funny business under the cover of a class which should essentially handle data mapping.

A l'occasion j'y mettrai bien un petit coup de refactoring. C'est quand même beaucoup mieux un code propre.

File: apps/backend/src/main/java/ch/heig/pdg/backend/presentation/AbstractController.java
Comment (noahboegli) [inline apps/backend/src/main/java/ch/heig/pdg/backend/presentation/AbstractController.java:1]: That is insane! Where did you learn all these objects-oriented techniques ?

File: apps/backend/src/main/java/ch/heig/pdg/backend/services/CategoryService.java [hunk 2]
```diff
@@ -13,26 +13,56 @@
...
 
     public CategoryService(CategoryRepository categoryRepository, InventoryRepository inventoryRepository, CategoryMapper categoryMapper) {
+        super(inventoryRepository);
         this.categoryRepository = categoryRepository;
-        this.inventoryRepository = inventoryRepository;
         this.categoryMapper = categoryMapper;
     }
 
+    public CategoryDTO getCategory(Integer id) {
+        return (CategoryDTO) this.categoryMapper.getDTO(
```
Comment (noahboegli) [inline apps/backend/src/main/java/ch/heig/pdg/backend/services/CategoryService.java:27]: Did you randomly press `Enter` on your keyboard ?

File: apps/backend/src/main/java/ch/heig/pdg/backend/services/CategoryService.java [hunk 3]
```diff
@@ -13,26 +13,56 @@
...
     }
+
+    public CategoryDTO addCategory(Integer inventoryId, CategoryDTO categoryDTO) {
+        Inventory inventory = this.checkInventory(inventoryId);
+
+        Category category = this.categoryMapper.createFromDTO(categoryDTO);
+        category.setInventory(inventory);
+
+        return (CategoryDTO) this.categoryMapper.getDTO(
+                this.categoryRepository.save(category)
```
Comment (noahboegli) [inline apps/backend/src/main/java/ch/heig/pdg/backend/services/CategoryService.java:49]: Again !

File: apps/backend/src/main/java/ch/heig/pdg/backend/services/InventoryService.java [hunk 4]
```diff
@@ -1,4 +1,59 @@
...
+                        this.inventoryMapper.updateFromDTO(inventory, inventoryDTO)
+                )
+        );
+    }
+
+    public InventoryDTO addInventory(InventoryDTO inventoryDTO) {
+        Inventory inventory = this.inventoryMapper.createFromDTO(inventoryDTO);
+
+        return (InventoryDTO) this.inventoryMapper.getDTO(
+                this.inventoryRepository.save(inventory)
```
Comment (noahboegli) [inline apps/backend/src/main/java/ch/heig/pdg/backend/services/InventoryService.java:50]: Do you happen to have a 5inch vertical monitor and lines can't be longer than 20 characters ?
It's Java not python !

File: apps/backend/src/main/java/ch/heig/pdg/backend/utils/DateFormatUtil.java
Comment (noahboegli) [inline apps/backend/src/main/java/ch/heig/pdg/backend/utils/DateFormatUtil.java:1]: Missing private constructor.

**Review by noahboegli — APPROVED** [review]
Thanks for the great work ! 

There's a few things that I'd change but let's go with that for now.

File: apps/backend/src/main/java/ch/heig/pdg/backend/dto/mapping/AbstractDataMapper.java
Comment (nigelmann) [reply apps/backend/src/main/java/ch/heig/pdg/backend/dto/mapping/AbstractDataMapper.java:1]: I understand but we can not re-create some entities without querying the repositories.
I prefer having a mapping layer that does all the mapping

File: apps/backend/src/main/java/ch/heig/pdg/backend/services/CategoryService.java (hunk 2 above)
Comment (nigelmann) [reply apps/backend/src/main/java/ch/heig/pdg/backend/services/CategoryService.java:27]: yes

File: apps/backend/src/main/java/ch/heig/pdg/backend/dto/mapping/AbstractDataMapper.java
Comment (noahboegli) [reply apps/backend/src/main/java/ch/heig/pdg/backend/dto/mapping/AbstractDataMapper.java:1]: Yeah but what I was saying is we put the requesting in an abstract service and we just pass the entity to the mapping ?

File: apps/backend/src/main/java/ch/heig/pdg/backend/utils/DateFormatUtil.java
Comment (nigelmann) [reply apps/backend/src/main/java/ch/heig/pdg/backend/utils/DateFormatUtil.java:1]: we can add it if you want, I don't care

**Pull request merged**